src/librustdoc/html/render.rs

```diff
@@ -1518,7 +1518,9 @@ impl Context {
                 if let Ok(mut redirect_out) = OpenOptions::new().create_new(true)
                                                                 .write(true)
                                                                 .open(&redir_dst) {
-                    try_err!(layout::redirect(&mut redirect_out, file_name), &redir_dst);
+                    buf.clear();
+                    try_err!(layout::redirect(&mut buf, file_name), &redir_dst);
+                    try_err!(redirect_out.write_all(&buf), &redir_dst);
```

Comment:
Seems like a BufWriter around redirect_out is what's really wanted here and below.

Reply:
This reuses the existing allocation of the buffer. An earlier (rebased-over) commit just wrapped the functions in `html/layout.rs` in BufWriters before i noticed that it was being handed Vecs most of the time anyway.

Reply:
By reusing the same `Vec` it minimizes the cost of allocation. Using a `BufWriter` every time you write to a file means having to allocate and deallocate a `BufWriter` each time which isn't super cheap. Considering the whole point of this is to shave off microseconds from an operation that can be repeated hundreds of thousands of times (in the literal sense, see winapi), reusing the `Vec` provides a small but important performance boost.

Reply:
Ah, didn't know we reused this thousands of times. `Vec` seems fine then, though ideally we'd add a `clear` to `BufWriter` probably...

Reply:
Right now the `Vec` is merely shared between the documentation on the item, and the redirect to the documentation, so the allocation isn't cached *across* items, but it's still better than a fresh allocation for writing the redirect. Also, there's no way as far as I can see to change what file a `BufWriter` is wrapping.

Reply:
*Technically* i think you could try `writer.flush(); *writer.get_mut() = &mut new_file;` to just swap out the inner writer, though i'm not sure if just having that flush there will save you from other weirdness. Also it requires that all the BufWriters be used with the same type.